src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs

```diff
@@ -530,5 +530,31 @@ public static string Page(this IUrlHelper urlHelper, string pageName, object val
                 host: host,
                 fragment: fragment);
         }
+
+        /// <summary>
+        /// Generates an absolute URL from the specified <paramref name="relativeUrl"/>
+        /// with the protocol and host name from the current request.
+        /// </summary>
+        /// <param name="helper">The <see cref="IUrlHelper"/>.</param>
+        /// <param name="relativeUrl">The relative URL.</param>
+        /// <returns>The generated absolute URL.</returns>
+        public static string Absolute(this IUrlHelper helper, string relativeUrl)
+        {
+            if (helper == null)
+            {
+                throw new ArgumentNullException(nameof(helper));
+            }
+
+            if (relativeUrl == null)
+            {
+                throw new ArgumentNullException(nameof(relativeUrl));
+            }
+
+            var httpContext = helper.ActionContext.HttpContext;
+            var protocol = string.IsNullOrEmpty(httpContext.Request.Scheme) ? "http" : httpContext.Request.Scheme;
+            var host = httpContext.Request.Host.ToUriComponent();
+            var absoluteUrl = protocol + "://" + host + relativeUrl;
```

Comment:
@dougbu not exactly sure how the `PathString` approach would work. The usage that was proposed and envisioned for this method looks like this:

```C#
Url.Absolute(Url.Action("Add", "Cart", new { productId = 1 }))
```

I did check and found that there is an implicit conversion from `string` to `PathString` but it doesn't work right with query params involved. I mean it encodes the `?` and so on and so forth.

Reply:
Why not add overloads of Action that give you an absolute value instead? E.g.
`Url.Action(absolute: true, "Add", "Cart", new { productId = 1 })`
or
`Url.ActionAbsolute("Add", "Cart", new { productId = 1 })`

Reply:
@Tratcher That was already discussed in the issue and rejected. The reason, I believe, being that it would result in too many overloads.

Reply:
Here's the relevant comment: https://github.com/aspnet/AspNetCore/issues/4890#issuecomment-432858777

Reply:
No validation, no PR acceptance. What validation bits do you mean here?

Reply:
httpContext.Request.Host is validated by the Server or host. https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-2.2#host-filtering

Reply:
@Tratcher the comment you linked to was discussing whether this issue should be addressed at all or not. And for the time being the issue was closed. But then it was reopened.

As for the specific implementation the discussion took place in previous PR which I now have trouble finding because it was in the old `aspnet/mvc` repo which is now archived...

Reply:
@Tratcher here found it:
https://github.com/aspnet/Mvc/pull/8532

And here's me double checking the specifics of the implementation with @pranavkm:

https://github.com/aspnet/Mvc/pull/8532#issuecomment-436766997

As far as I can tell he is the part of the team.

Reply:
I'll follow up with him, the design still needs work.

Reply:
Sorry if my comments gave off the wrong impression, I just thought it was a neat idea to avoid having to end up with lots of `Absolute~` methods – I do get that this makes it a bit more difficult to implement since you can’t trust the parameter to be a result from another URLHelper method.

Reply:
@poke You're good. The idea was appealing in it's simplicity, but it's on @pranavkm to work through the scenarios.

Reply:
@TKharaishvili \ @poke we had an offline chat about this and the fact that the API accepts an arbitrary string does make it difficult to reason about it's origin and consequently things like what to do with encoding. We considered documenting it to say the only inputs that should be passed in to would be values returned by other `IUrlHelper` methods, but that isn't a sufficient safeguard against improper use of the API which is problematic given how easily accessible `IUrlHelper` is within MVC. The plan is to go back to the API suggested in the original issue. My apologies for the confusion here.

Reply:
Couldn't the method check for a leading slash and then properly document that the passed (relative) URL will be treated as root-relative?

That should make it safe enough while also avoiding surprises from the usage.

Reply:
@poke that wouldn't address the concern about encoding the passed `string`.

Reply:
@dougbu Meh, good point… 🙁

Reply:
@pranavkm so what's the idea here? Should we implement `Url.AbsoluteAction`, `Url.AbsoluteRouteUrl` and `Url.AbsolutePage`? And for all the overloads? Or should we ease into it and just do `Url.AbsoluteAction` first?

Also, should we move this bit of discussion back to the issue itself? Cause when the important part of the conversation took place in the PR before, it made it harder to follow it some time later.

And should I close this PR and open a new one? Or should I just update this one?